### PR TITLE
Attention Refactor (WIP)

### PR DESCRIPTION
**Pull request by stefanradev93**
This PR refactors the transformers module for internal consistency and directly exposes `attention_mask` and `use_causal_mask` in the relevant transformers. The following changes were made:

- Transformer building blocks were moved into a dedicated `attention` module.
- Abstract base class `Transformer` was added to easily tell apart transformer summaries from other summary nets.
- Files were renamed to reflect semantic names.
- `TimeSeriesTransformer` can now act as a many-to-many network (e.g., for modeling time-varying targets)

It also prepares to address #https://github.com/bayesflow-org/bayesflow/issues/626.

@arrjon @paul-buerkner It remains to decide how we want the attention mask passed? Should we search for it in the simulator outputs (as we do for other special args like `sample_weights`)? This would have the advantage that the mask can be constructed very flexibly.

### CODE REVIEW

**Comment by paul-buerkner** [comment]
Nice! Yes, I would love to search for `attention_mask` in the simulator output! In fact, I cannot really think of an alternative that would be nearly as nice. Is there any disadvantage to doing so?

**Comment by stefanradev93** [comment]
Not any I could think of, except some overhead on the user's side which may be unavoidable. @arrjon if you concur, I will proceed with adding the search for attention mask among simulation outputs.

**Comment by arrjon** [comment]
Sounds good to me as it gives maximal flexibility to the user!

**Comment by stefanradev93** [comment]
I now allowed for `mask` and `attention_mask` keyword arguments in `compute_metrics` of the `approximator`, which are expected to be part of the output (or `None` - default) and be propagated both to the inference and summary network, depending on the signature of each network. 

The case for sampling is tricky tho, as now the user will have to provide the `attention_mask` or `mask` arguments as `**kwargs` instead of passing them as part of the `conditions` dict. This may pose some confusion, but expecting the masks to be part of the conditions invalidates our internal logic...

**Comment by paul-buerkner** [comment]
Can you elaborate why we cannot use `conditions` for this purpose?

**Comment by stefanradev93** [comment]
I can solve it with a bunch of checks for now.

**Comment by paul-buerkner** [comment]
I don’t fully understand what you mean by this. Solve what? Making it part
of conditions?

Stefan Radev ***@***.***> schrieb am Di. 3. Feb. 2026 um
14:56:

**Comment by stefanradev93** [comment]
@arrjon Can you please check if my latest commit enables the functionality you needed?

**Comment by arrjon** [comment]
Thanks @stefanradev93, now the masking seems to work!

**Comment by stefanradev93** [comment]
@LarsKue @arrjon  I would actually like to discard that PR in favor of the approximator refactor https://github.com/bayesflow-org/bayesflow/pull/633, which would make mask handling much easier. Thus, I would only keep the attention module refactor from this branch/PR.

**Comment by LarsKue** [comment]
@stefanradev93 so this can be closed?
